jbpm-recommendation-spark-random-forest/src/main/java/org/jbpm/prediction/randomforest/SparkRandomForest.java

```diff
@@ -0,0 +1,213 @@
+                        DataTypes.createStructField("label", DataTypes.IntegerType, false),
+                });
+
+        predictionSchema = DataTypes.createStructType(
+                new StructField[] {
+                        DataTypes.createStructField("user", DataTypes.StringType, false),
+                        DataTypes.createStructField("level", DataTypes.IntegerType, false)
+                });
+
+        // Create an empty dataframe to store incoming data
+        df = session.createDataFrame(jsc.emptyRDD(), trainingSchema).toDF();
+
+        final StringIndexer userIndexer = new StringIndexer().setInputCol("user").setOutputCol("userIndex");
+        final OneHotEncoder userEncoder = new OneHotEncoder().setInputCol("userIndex").setOutputCol("userVec");
+
+        final VectorAssembler assembler = new VectorAssembler().setInputCols(new String[]{"userVec", "level"}).setOutputCol("features");
+        final RandomForestClassifier randomForest = new RandomForestClassifier().setNumTrees(this.numTrees)
+                .setMaxBins(32)
+                .setMaxMemoryInMB(128)
+                .setMaxDepth(3);
+
+        // Pipeline stages consist of: indexing users -> encoding features -> assembling vector -> random forest
+        final PipelineStage[] stages = new PipelineStage[]{
+                userIndexer, userEncoder, assembler, randomForest
+        };
+
+        pipeline = new Pipeline().setStages(stages);
+    }
+
+
+    /**
+     * Add the data provided as a map to a Smile {@link smile.data.Dataset}.
+     *
+     * @param data    A map containing the input attribute names as keys and the attribute values as values.
+     * @param outcome The value of the outcome (output data).
+     */
+    public void addData(Map<String, Object> data, Object outcome) {
+        // convert input data to a Row
+        final List<Row> rows = ImmutableList.of(
+                RowFactory.create(
+                        data.get("ActorId").toString(),
+                        data.get("level"),
+                        ((Boolean) outcome) ? 1 : 0
+                ));
+        // Append the row to the existing dataframe
+        df = df.union(session.createDataFrame(rows, trainingSchema));
+    }
+
+    /**
+     * Returns the service's identifier
+     *
+     * @return The service identifier
+     */
+    @Override
+    public String getIdentifier() {
+        return IDENTIFIER;
+    }
+
+    /**
+     * Returns a model prediction given the input data
+     *
+     * @param task      Human task data
+     * @param inputData A map containing the input attribute names as keys and the attribute values as values.
+     * @return A {@link PredictionOutcome} containing the model's prediction for the input data.
+     */
+    @Override
+    public PredictionOutcome predict(Task task, Map<String, Object> inputData) {
+        Map<String, Object> outcomes = new HashMap<>();
+
+        if (pm != null) {
```

Comment:
You may change the condition to positive and return immediately the new PredictionOutcome, without 'else' clause and one level less of indentation.
```suggestion
        if (pm == null)
            return new PredictionOutcome();
```